### ISSUE TITLE
Is it possible to support QMK firmware?

### ISSUE THREAD

**Issue by acme64** [antonpup/Aurora#1344]
Forgive me if i have a term wrong or im asking for the wrong thing, i'm not well versed with this. I'm wondering if it'd be possible to support QMK firmware a lot of custom keyboards use? here's their project if you're not familiar https://github.com/qmk/qmk_firmware

**Comment by clrxbl** [comment]
I don't think you can modify QMK lighting without re-flashing it every single time.

**Comment by gitmacer** [comment]
In theory it is possible to communicate over USB serial with the keyboard but I do not know if there are firmware limitations. You should ask for usb serial rgb controll in the firmware repo.

**Comment by helluvamatt** [comment]
@acme64 Don't despair! It's possible, and not even that difficult!

@gitmacer I'm very familiar with the QMK firmware. There's been some development of a "RAW HID" protocol for keyboards that support an "RGB Matrix" component (basically, any keyboard with several addressable RGB leds can support the RGB matrix, but it's targeted at underglow and back-light RGB). I've done some work getting RGB Matrix to work with the [Massdrop CTRL keyboard](https://www.massdrop.com/buy/massdrop-ctrl-mechanical-keyboard), which includes 119 RGB leds (1 under each of the 87 keys and 32 more around the side of the keyboard). The code on the PC side is actually extremely simple on Windows. My code has been mostly .NET P/Invoke calls directly to kernel32.dll, hal.dll, or usb.dll. I would recommend against a CDC endpoint in QMK since we tend to keep CDC for debugging (printf in our firmware generally routes to the CDC endpoint so anyone listening on our COM port gets the printf calls printed there.)

**Comment by helluvamatt** [comment]
Unfortunately, my Massdrop CTRL developed a fault and had to be sent back for replacement.

When the replacement arrives, I'll see about getting an adapter written for Aurora to talk to my HID RAW protocol.

If anyone would like to have a look at the protocol from the firmware side, you can have a look at [ctrl.c](https://github.com/helluvamatt/qmk_firmware/blob/ctrl_matt/keyboards/massdrop/ctrl/ctrl.c)

**Comment by gitmacer** [comment]
@helluvamatt first of all thank you for your help! We really appreciate that.
But I think you did not fully understand our way to implement devices please have a look at these files. I think Atmoorb and Wooting are easy to understand ones: [here](https://github.com/antonpup/Aurora/tree/master/Project-Aurora/Project-Aurora/Devices). 
The wrapper files are used to convert light signals from games that support "razer chroma", "lightfx" or other. If you have any additional question and want a fast answer you can always conntact us over Discord [here](https://discord.gg/YAuBmg9).

**Comment by helluvamatt** [comment]
@gitmacer Thanks for pointing me in the right direction!

It looks like the [SteelSeriesHID](https://github.com/antonpup/Aurora/blob/master/Project-Aurora/Project-Aurora/Devices/SteelSeriesHID/SteelSeriesHID.cs) device is going to be fairly close in that we are exposing a "vendor-defined HID class device" for host programs (such as Aurora to communicate with).

**Comment by daniel5151** [comment]
@helluvamatt I'm planning on looking into this feature soonish, and I was hoping you could upload any preliminary code you've got so-far. I'm sure I could figure stuff out myself, but it always helps to have a jumping-off point :)

**Comment by helluvamatt** [comment]
@daniel5151 I’m afraid not at the moment.

**Comment by daniel5151** [comment]
Oof. No problem tho. If you reconsider, let me know :)

**Comment by eaaaaaaast** [comment]
@helluvamatt I'd also be willing to throw down on a bounty to get this done for the CTRL.

**Comment by acme64** [comment]
So, anyone know whats up with this? i saw some sort of support was added to the QMK git, but i have no idea what that does?

**Comment by TheNinty** [comment]
I was thinking of building a QMK keyboard. I would love to have aurora working on my own customized keyboard, so does anyone know if this is still being worked on?

**Comment by Ming-A** [comment]
so i just built a QMK keyboard and would like to know if its possible?

**Comment by diogotr7** [comment]
The best chance of having qmk support is through OpenRGB. Follow the relevant issue here: https://gitlab.com/CalcProgrammer1/OpenRGB/-/issues/633

As far as i know, this fork does have support for a massdrop keyboard, although it does require enabling some setting in the firmware: https://gitlab.com/Kasper24/OpenRGB/-/tree/qmk_rgb_matrix

I'm not very familiar with how the QMK support works but if you want more info join the OpenRGB discord.

**Comment by tirrorex** [comment]
Thanks for the links @diogotr7, so in theory what i should do is making kasper fork compatible with my keychron and then interface it with aurora yes? (for razer chroma support).

Am i missing some things or is it all that i need to do?
